northstar/src/api/codec.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2020 ESRLabs
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+use super::model;
+use bytes::{Buf, BufMut};
+use futures::Stream;
+use std::{
+    cmp::min,
+    io::ErrorKind,
+    pin::Pin,
+    task::{self, Poll},
+};
+use task::Context;
+use tokio::io::{self, AsyncRead, AsyncWrite};
+use tokio_util::codec::{Decoder, Encoder};
+
+/// Newline delimited json codec for api::Message that on top implementes AsyncRead and Write
+pub struct Framed<T> {
+    framed: tokio_util::codec::Framed<T, Codec>,
+}
+
+pub fn framed<T: AsyncRead + AsyncWrite>(io: T) -> Framed<T> {
+    Framed {
+        framed: tokio_util::codec::Framed::new(io, Codec {}),
+    }
+}
+
+/// Newline delimited json
+struct Codec;
+
+impl Decoder for Codec {
+    type Item = model::Message;
+    type Error = io::Error;
+
+    fn decode(&mut self, src: &mut bytes::BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        if let Some(position) = src.iter().position(|b| *b == b'\n') {
+            let buf = src.split_to(position);
+            // Consume the newline
+            src.advance(1);
+            match serde_json::from_slice::<model::Message>(&buf) {
+                Ok(message) => Ok(Some(message)),
+                Err(e) => Err(io::Error::new(ErrorKind::Other, e)),
```

Comment:
better use `ErrorKind::InvalidData`

Reply:
With `ErrorKind::Invalid` we'd discard the error information from `serde`.

Reply:
why is this?
`Err(e) => Err(io::Error::new(ErrorKind::InvalidData, e)),`
what information is lost here?

Reply:
Ah. Its a `ErrorKind`. My bad. This makes sense. I'll change that.